src/unistd.rs

```diff
@@ -2136,3 +2146,424 @@ mod setres {
+    Name(String),
+    UidWithBufsize(Uid, usize),
+    NameWithBufsize(String, usize)
+}
+
+/// Query a group. 
+///
+/// The buffer size variants should not be needed 99% of the time; the default buffer size of 1024
+/// should be more than adequate. Only use the buffer size variants if you receive an ERANGE error
+/// without them, or you know that you will be likely to.
+///
+/// # Examples
+///
+/// ```ignore
+/// use nix::unistd::{Group, GroupQuery};
+/// // On many systems there's no `root` group; on FreeBSD for example it's known as `wheel`
+/// let find = String::from("root");
+/// // Returns an Option<Result<Group>>, thus the double unwrap. Will panic if there's no `root`
+/// // group.
+/// let res = Group::query( GroupQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+#[derive(Debug)]
+pub enum GroupQuery {
+    /// Get a group by GID. 
+    ///
+    /// Internally, this function calls
+    /// [getgrgid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Gid(Gid),
+    /// Get a group by name. 
+    ///
+    /// Internally, this function calls
+    /// [getgrnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    GidWithBufsize(Gid, usize),
+    NameWithBufsize(String, usize)
+}
+
+pub trait Queryable<Q> where Self: ::std::marker::Sized {
+    fn query(q: Q) -> Option<Result<Self>>;
+}
+
+/// Default buffer size for system user and group querying functions
+const PWGRP_BUFSIZE: usize = 1024;
+
+impl Queryable<UserQuery> for User {
+    fn query(q: UserQuery) -> Option<Result<Self>> {
+        let mut cbuf = match q {
+            UserQuery::UidWithBufsize(_, size) | 
+            UserQuery::NameWithBufsize(_, size) => {
+                vec![0 as c_char; size]
+            },
+            _ => {
+                vec![0 as c_char; PWGRP_BUFSIZE]
+            }
+        };
+
+        let mut pwd: libc::passwd =  unsafe { mem::zeroed() };
+        let mut res = ptr::null_mut();
+
+        let i = { 
+            unsafe { Errno::clear(); }
+
+            match q {
+                UserQuery::UidWithBufsize(Uid(uid), _) |
+                UserQuery::Uid(Uid(uid)) => {
+                    unsafe {
+                        libc::getpwuid_r(uid, &mut pwd,
+                                         cbuf.as_mut_ptr(),
+                                         cbuf.len(), &mut res) 
+                    }
+                },
+                UserQuery::NameWithBufsize(name, _) |
+                UserQuery::Name(name) => {
+                    unsafe {
+                        libc::getpwnam_r(CString::new(name).unwrap().as_ptr(), 
+                                         &mut pwd, cbuf.as_mut_ptr(), cbuf.len(), &mut res) 
+                    }
+                },
+            }
+
+        };
+
+        match i {
+            0 => {
```

Comment:
If you're just doing an if/else like this I'd recommend doing that instead of a match, which is a little more heavyweight.

Reply:
changed it to an `if/else`.